### DISCUSSION TITLE
How to obtain pointer to data in nb::bytes object?

### DISCUSSION THREAD

**Discussion by jeffective** [wjakob/nanobind#568] [Q&A]
I am wrapping this function:

```c
int ecx_SDOwrite(ecx_contextt *context, uint16 Slave, uint16 Index, uint8 SubIndex,
                 boolean CA, int psize, const void *p, int Timeout);
```

How do I get a `void*` to the data in a `nb::bytes` object?

```cpp
// p is a void pointer to the memory storing the bytes
// how do i obtain a pointer to the data in nb::bytes ?

auto SDOwrite(uint16 Slave, uint16 Index, uint8 SubIndex, boolean CA, nb::bytes data, int Timeout_us){
        
        // how do i get p? 
        return ecx_SDOwrite(&this->context, Slave, Index, SubIndex, CA, data.size(), p, Timeout_us);
    }
```

**Comment by jeffective** [comment]
i guess i could cast to a c string but then I have an extra null byte

**Reply by jeffective** [reply]
and the bytes cannot contain any null bytes

**Reply by nurpax** [reply]
I think it's ok for bytes to contain \0 chars in Python.  I'm passing bytes values to through to my bindings and it works.  This was changed recently:

https://github.com/wjakob/nanobind/commit/18376ff56d35c3adf98bc7ba77799fe6ca33a20b

I didn't check how to go the other way though.

**Comment by jeffective** [comment]
Going to use a bind_vector of uint8_t instead and access it with `.data()` and `.size()`
https://nanobind.readthedocs.io/en/latest/api_extra.html#stl-vector-bindings
